ast/src/analyzed/display.rs

```diff
@@ -31,27 +31,43 @@ impl Display for DegreeRange {
 
 impl<T: Display> Display for Analyzed<T> {
     fn fmt(&self, f: &mut Formatter<'_>) -> Result {
-        let (mut current_namespace, mut current_degree) = (AbsoluteSymbolPath::default(), None);
-        let mut update_namespace =
-            |name: &str, degree: Option<DegreeRange>, f: &mut Formatter<'_>| {
-                let mut namespace =
-                    AbsoluteSymbolPath::default().join(SymbolPath::from_str(name).unwrap());
-                let name = namespace.pop().unwrap();
-                if namespace != current_namespace {
-                    current_namespace = namespace;
-                    current_degree = degree;
-                    writeln!(
-                        f,
-                        "namespace {}{};",
-                        current_namespace.relative_to(&Default::default()),
-                        degree.map(|d| format!("({d})")).unwrap_or_default()
-                    )?;
-                } else {
-                    // If we're in the same namespace, the degree must match
-                    assert_eq!(current_degree, degree);
-                };
-                Ok((name, !current_namespace.is_empty()))
+        // Compute map from namespace name to its degree (if not None).
+        let namespace_degrees = self
+            .definitions
+            .values()
+            .filter_map(|(symbol, _)| {
+                symbol.degree.map(|degree| {
+                    let (namespace, _) = split_into_namespace_and_name(&symbol.absolute_name);
+                    (namespace, degree)
+                })
+            })
+            .unique()
+            .fold(BTreeMap::default(), |mut degrees, (namespace, degree)| {
+                let inserted = degrees.insert(namespace.clone(), degree).is_none();
```

Comment:
```suggestion
                let is_new = degrees.insert(namespace.clone(), degree).is_none();
```
It's always inserted, right? Could also be inlined inside the `assert` IMO.